_posts/2015-11-16-behat-to-the-fixture.md

```diff
@@ -0,0 +1,272 @@
+title: "Behat to the Fixture"
+excerpt: "How to use Behat to write Fixtures in Symfony2"
+modified: 2015-08-14
+tags: [behat, symfony2, fixtures]
+comments: true
+image:
+  feature: behat-to-the-fixture.jpg
+  credit: Alban Pommeret
+  creditlink: http://reputationvip.io
+---
+
+## What is a fixture ?
+
+[DoctrineFixturesBundle documentation](http://symfony.com/doc/current/bundles/DoctrineFixturesBundle/index.html) :
+
+>Fixtures are used to load a controlled set of data into a database. This data can be used for testing or could be the initial data required for the application to run smoothly. Symfony2 has no built in way to manage fixtures but Doctrine2 has a library to help you write fixtures.
+
+The goal of this article is to propose a new method based on Gherkin language and behat 3 to write and load fixtures on Symfony2.
+
+For further information about behat test, refer to this article : [BDD, Behaviour Driven Development](/bdd-behat-driven-development) 
+
+***
+
+## Current ways to write fixtures on Symfony2
+
+There are two main ways to write symfony2 fixtures.
+
+Write your entities directly in PHP :
+
+{% highlight php startinline=true %}
+class LoadUserAndBookData implements FixtureInterface
+{
+    public function load(ObjectManager $manager)
+    {
+         $userJames = new User();
+         $userJames->setName('james');
+         $userJames->setPassword('pwd');
+         $manager->persist($userJames);
+     
+         $userJohn = new User();
+         $userJohn->setName('john');
+         $userJohn->setPassword('pwd'); 
+         $manager->persist($userJohn); 
+     
+         $book1 = new Book();
+         $book1->setTitle('Lord of the ring');
+         $book1->setOwner($userJames); 
+         $manager->persist($book1);
+     
+         $book2 = new Book();
+         $book2->setTitle('Harry Potter 1');
+         $book2->setOwner($userJohn); 
+         $manager->persist($book2);
+     
+         $manager->flush();
+    }
+}
+{% endhighlight %}
+
+
+Or use yml format (with hautelook/alice-bundle) :
+
+{% highlight gherkin %}
+AppBundle\Entity\User:
+    user1: 
+        name: james
+        password: pwd
+    user2:
+        name: john
+        password: pwd
+
+AppBundle\Entity\Book:
+    book1: 
+        title: Lord of the ring
+        owner: @user1
+    book2:
+        title: Harry Potter 1
+        owner: @user2
+{% endhighlight %}
+
+
+## Fixtures in Gherkin language
+
+I propose here a third way, based on Gherkin language and interpreted by behat :
+
+{% highlight gherkin %}
+Feature: Load data.
+
+    Scenario: fixture 1
+        Given users :
+            | name  | password |
+            | james | pwd      |
+            | john  | pwd      |
+        Given books :
+            | title            | user  |
+            | Lord of the ring | james |
+            | Harry Potter 1   | john  |
+
+{% endhighlight %}
+
+Gherkin language is more readable than yml format. Even if you are not a developer, you can write it.
```

Comment:
This is however one of the main arguments for YML syntax. Human readibility and writability.